syntax = "proto3";
package prodvana.desired_state.model;
option go_package = "github.com/prodvana/prodvana-public/go/prodvana-sdk/proto/prodvana/desired_state/model";
import "prodvana/desired_state/model/desired_state.proto";
import "google/protobuf/timestamp.proto";

message Entity {
    Identifier id = 1;
    string desired_state_id = 2;
    string root_desired_state_id = 3;
    Annotations annotations = 4;
    Status status = 5;
    SimpleStatus simple_status = 17;
    State starting_state = 6;
    State last_seen_state = 7;
    State desired_state = 8;
    repeated ConditionState precondition_statuses = 9;
    repeated StatusExplanation status_explanations = 10;
    repeated DebugLog logs = 11;
    ActionExplanation action_explanation = 12;
    google.protobuf.Timestamp last_update_timestamp = 13;  // when the entity state last changed
    google.protobuf.Timestamp last_fetched_timestamp = 14;  // when prodvana last fetched entity state, best effort
    google.protobuf.Timestamp last_applied_timestamp = 15;  // when prodvana last applied state for this entity, best effort

    repeated Identifier dependencies = 16;
    Lifecycle lifecycle = 18;
    // next tag: 19
}

message EntityGraph {
    Identifier root = 1;
    repeated Entity entities = 2;
}
